### PR TITLE
[DO NOT MERGE] increase the on-wire precision of time

### LINKED ISSUE #15262: The  second-level on-wire precision of unversioned.Time is too low
When marshing unversioned.Time to JSON/YAML, the precision decreases to second-level, which is too low, i.e.,
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/time.go#L105
```
json.Marshal(t.UTC().Format(time.RFC3339))
```

We need to use RFC3339Nano instead, which keeps the nanosecond-level precision.

@bgrant0607 @saad-ali

### PR DESCRIPTION
**Pull request by caesarxuchao**
fix #15262

This PR uses RFC3339Nano instead of RFC3339 when marshaling/unmarshaling unversioned.Time. This increase the on-wire precision of time to nanosecond-level.

This should be a backward-compatible change, because we can use RFC3339 to parse a RFC3339Nano formatted string, and vice-vesa.  (see http://play.golang.org/p/GfNp7xAjNv)

@saad-ali @bgrant0607

### CODE REVIEW

**Comment by bgrant0607** [comment]
Tip: It's possible to assign the PR when creating the PR.

File: pkg/api/unversioned/time.go [hunk 1]
```diff
@@ -76,6 +76,12 @@ func (t Time) Rfc3339Copy() Time {
 	return Time{copied}
```
Comment (saad-ali) [inline pkg/api/unversioned/time.go]: Is `Rfc3339Copy()` still needed?

Comment (caesarxuchao) [inline pkg/api/unversioned/time.go]: It's not being used by anyone, but I feel it would be useful if someone want to get second-level copies. Your call :)

Comment (saad-ali) [inline pkg/api/unversioned/time.go]: Let's remove it, since it is unused. It can be added back in if someone needs it.

Comment (caesarxuchao) [inline pkg/api/unversioned/time.go]: Removed.

**Comment by saad-ali** [comment]
LGTM

**Comment by saad-ali** [comment]
Please squash your commits.

**Comment by caesarxuchao** [comment]
Squashed.

**Comment by caesarxuchao** [comment]
remove the LGTM to allow @lavalamp to take a look.

**Comment by lavalamp** [comment]
It sounds like this is a non-breaking change for *go* clients. What about other clients, though?

**Comment by lavalamp** [comment]
@pmorie Can you remember the reason why we serialized the level of precision we did?

**Comment by caesarxuchao** [comment]
Yeah, it would break non-go client if they expect to see time in RFC3339 format.

**Comment by pmorie** [comment]
@lavalamp I'm not sure it was intentional but ntp only supports
ms-resolution so unless someone is using more precise network time nanos
are not going to be meaningful

**Comment by caesarxuchao** [comment]
@pmorie, yes, it's not meaningful accross machines. However, it's meaningful if we are comparing the time stamps generated by the same machine, e.g., the timestamps in Events that are generated by the same component.

**Comment by pmorie** [comment]
That's true @caesarxuchao.

File: pkg/api/unversioned/time.go (hunk 1 above)
Comment (pmorie) [inline pkg/api/unversioned/time.go]: +1 to removing this, YAGNI

**Comment by pmorie** [comment]
@caesarxuchao How about adding a note in the appropriate types.go files describing the serialization precision of `ObjectMeta.CreationTimestamp`?

**Comment by caesarxuchao** [comment]
@pmorie I fixed several occurrences of "RFC3339" in types.go and regenerated the swagger spec. PTAL. Thanks.

**Comment by caesarxuchao** [comment]
@bgrant0607, this will be a breaking change for *non-go* clients. Shall we proceed?

**Comment by pmorie** [comment]
LGTM, @lavalamp ?

**Comment by lavalamp** [comment]
I think this is a breaking API change. I'm not sure if it's an important breaking change, though. Let's let @bgrant0607 weigh in.

**Comment by pmorie** [comment]
Good point @lavalamp

**Comment by smarterclayton** [comment]
I suspect this will break most naive date parsing in client code - have we assessed its impact on naive Ruby, Python, and JavaScript libraries that would be used to parse a date from json?

**Comment by smarterclayton** [comment]
And Java

**Comment by bgrant0607** [comment]
Let's hold off for now, until we assess the impact on the other client languages mentioned.

**Pull request closed without merging**